add core.redux-provider extension and support

christianvogt opened the pull request:
fixes: https://issues.redhat.com/browse/HAC-1296

New extension `core.redux-provider`.
Provides a configuration for establishing a Redux instance scoped for the contributing plugin.

`ReduxExtensionProvider` component may be used by applications to support redux providers at the root.

Each redux provider will be scoped by the extension with a unique store and context.

Usage of redux apis must supply the context to gain access to the respective store. eg `createSelectorHook(myContext);`

vojtechszocs commented on packages/lib-utils/src/app/redux/ReduxExtensionProvider.tsx:
```diff
@@ -0,0 +1,27 @@
+        children,
+      )}
+    </>
+  ) : null;
```
There are currently two scenarios which may cause extensions to be put into or out of use at any point in time:
- change in a plugin's enabled status
- change in the feature flags, which in turn impacts the availability of individual extensions

Extensions in use are computed as all extensions from plugins that are enabled, which also meet the feature flag criteria according to the current feature flag values.

Therefore,
```ts
useResolvedExtensions<ReduxProvider>(isReduxProvider)
```
may cause this hook to be re-run multiple times, providing different results on each run.

I think we need to guard against this, otherwise we may "un-render" and then "re-render" a Redux `Provider` which would effectively reset it (?)

christianvogt replied:
The context and store should be the same instance on a re-render. So even if the provider was re-rendered, those who rely on the redux store would still be accessing the same redux context and store.

vojtechszocs replied:
One more question, if a `ReduxProvider` extension is put out of use (due to change in flags or due to the owner plugin getting disabled) but some component down in the tree still attempts to access the associated context or store, the responsibility to handle this is on the consuming component, correct?

christianvogt replied:
Correct. If a nested component relies on redux and the provider is not present, it's their responsibility. Under normal circumstances the system will ensure its presence in time for consumption.

A redux provider that is disabled by a flag, while possible, this scenario is less likely to happen as the presence would have a no-op effect if unused and therefore using a flag wouldn't have any real benefit.

If the plugin is disabled, then all contributions from said plugin would also be disabled and therefore there would be no reliance on the provider.

vojtechszocs commented on packages/lib-utils/src/app/redux/ReduxExtensionProvider.test.tsx:
```diff
@@ -0,0 +1,118 @@
+
+describe('ReduxExtensionProvider', () => {
+  afterEach(() => {
+    jest.clearAllMocks();
```
I'd suggest to do any mock clear or reset in `beforeEach` instead of `afterEach`.

Note that we already have
```js
clearMocks: true,
```
in `packages/common/test-configs/jest-config-base.ts` file.

I personally prefer to be specific about how and when a test suite clears or resets its mocks, like it's done above, instead of a global `clearMocks: true` setting (which I'd like to remove in the future).

vojtechszocs commented on packages/lib-utils/src/app/redux/ReduxExtensionProvider.test.tsx:
```diff
@@ -0,0 +1,118 @@
+      </ReduxExtensionProvider>,
+    );
+
+    expect(container.firstElementChild?.tagName).toBe('SPAN');
```
Just wondering, does the tag name have to be uppercase?

christianvogt replied:
yes, `tagName` is always upper case

christianvogt commented on packages/lib-utils/src/app/redux/ReduxExtensionProvider.tsx:
```diff
@@ -0,0 +1,27 @@
+import { Provider } from 'react-redux';
+
+/**
+ * Consumes the core.redux-
```
whoops, i guess i never finished my thought lol

florkbr replied:
@christianvogt FYI @vojtechszocs is OOO for the remainder of the week. Were you going to address these comments? I can re-review so we can try to get this in.

christianvogt commented on packages/lib-utils/src/app/redux/ReduxExtensionProvider.tsx:
```diff
@@ -0,0 +1,27 @@
+/**
+ * Consumes the core.redux-
+ */
+const ReduxExtensionProvider: React.FC = ({ children }) => {
```
Is that a convention we're following? I like provider because those who use redux are familiar with the setup component named `Provider`.

vojtechszocs replied:
Not a convention, I'm OK with the current name - just realized that the `Provider` suffix means this component renders a Redux `Provider` component for each extension instance.

(I've based my original suggestion around the fact this component handles/interprets `ReduxProvider` extensions, but I agree that `Provider` suffix sounds more familiar to those already using Redux.)

christianvogt commented:
Addressed review comments except the renaming of `ReduxExtensionProvider`. Unless there's a convention we're following, the name currently aligns with redux api.

vojtechszocs commented:
/lgtm

florkbr commented:
/approve
